File: backend/app/services/memory_normalize.py
"""项目设定结构化数据的规范化（读写兼容旧格式）"""
from __future__ import annotations

import json
from typing import Any, Dict, List


DEFAULT_WORLD_CATEGORIES = [
    "时代背景",
    "地理环境",
    "力量/规则",
    "社会结构",
    "文化习俗",
    "其他设定",
]

CHAR_ROLES = ["主角", "重要配角", "反派", "配角", "其他"]


def _as_dict(raw: Any) -> Any:
    if isinstance(raw, str):
        s = raw.strip()
        if not s:
            return None
        if s.startswith("{") or s.startswith("["):
            try:
                return json.loads(s)
            except json.JSONDecodeError:
                return raw
    return raw


def normalize_storyline(raw: Any) -> Dict[str, Any]:
    raw = _as_dict(raw)
    if raw is None or raw == "":
        return {"summary": "", "stages": []}
    if isinstance(raw, str):
        return {"summary": raw, "stages": []}
    if isinstance(raw, list):
        stages = [_normalize_stage(s) for s in raw]
        return {"summary": "", "stages": stages}
    if isinstance(raw, dict):
        stages_raw = raw.get("stages") or raw.get("nodes") or []
        if not isinstance(stages_raw, list):
            stages_raw = []
        return {
            "summary": str(raw.get("summary") or raw.get("overview") or ""),
            "stages": [_normalize_stage(s) for s in stages_raw],
        }
    return {"summary": "", "stages": []}


def _normalize_stage(item: Any) -> Dict[str, str]:
    if isinstance(item, str):
        return {"title": item[:40], "summary": item}
    if isinstance(item, dict):
        title = str(item.get("title") or item.get("name") or "")
        summary = str(
            item.get("summary")
            or item.get("description")
            or item.get("content")
            or ""
        )
        return {"title": title, "summary": summary}
    return {"title": "", "summary": ""}


def normalize_key_points(raw: Any) -> List[Dict[str, str]]:
    raw = _as_dict(raw)
    if not raw:
        return []
    if not isinstance(raw, list):
        return []
    result = []
    for item in raw:
        if isinstance(item, str):
            result.append({"title": item[:40] if len(item) > 40 else item, "summary": item})
        elif isinstance(item, dict):
            title = str(item.get("title") or item.get("name") or "")
            summary = str(
                item.get("summary")
                or item.get("description")
                or item.get("content")
                or item.get("point")
                or ""
            )
            if not title and summary:
                title = summary[:40]
            result.append({"title": title, "summary": summary})
    return result


def normalize_character(raw: Any) -> Dict[str, Any]:
    if not isinstance(raw, dict):
        return {
            "name": "",
            "description": "",
            "personality": "",
            "background": "",
            "goals": "",
            "role": "配角",
            "avatar": "",
            "color": "",
        }
    return {
        "name": str(raw.get("name") or ""),
        "description": str(raw.get("description") or ""),
        "personality": str(raw.get("personality") or ""),
        "background": str(raw.get("background") or ""),
        "goals": str(raw.get("goals") or ""),
        "role": str(raw.get("role") or "配角"),
        "avatar": str(raw.get("avatar") or ""),
        "color": str(raw.get("color") or ""),
    }


def normalize_characters(raw: Any) -> List[Dict[str, Any]]:
    raw = _as_dict(raw)
    if not isinstance(raw, list):
        return []
    return [normalize_character(c) for c in raw]


def normalize_world_building(raw: Any) -> Dict[str, Any]:
    """统一为 { categories: [{ name, items: [{ title, content }] }] }"""
    raw = _as_dict(raw)
    if not raw:
        return {"categories": [{"name": n, "items": []} for n in DEFAULT_WORLD_CATEGORIES]}

    if isinstance(raw, dict) and isinstance(raw.get("categories"), list):
        cats = []
        for c in raw["categories"]:
            if not isinstance(c, dict):
                continue
            name = str(c.get("name") or c.get("title") or "未命名分类")
            items_raw = c.get("items") or []
            items = []
            if isinstance(items_raw, list):
                for it in items_raw:
                    if isinstance(it, str):
                        items.append({"title": it[:30], "content": it})
                    elif isinstance(it, dict):
                        items.append({
                            "title": str(it.get("title") or it.get("key") or it.get("name") or ""),
                            "content": str(it.get("content") or it.get("value") or it.get("description") or ""),
                        })
            cats.append({"name": name, "items": items})
        if not cats:
            cats = [{"name": n, "items": []} for n in DEFAULT_WORLD_CATEGORIES]
        return {"categories": cats}

    # 旧版扁平 key: value
    if isinstance(raw, dict):
        items = []
        for k, v in raw.items():
            if k in ("categories",):
                continue
            items.append({"title": str(k), "content": "" if v is None else str(v)})
        return {
            "categories": [
                {"name": "综合设定", "items": items},
                *[{"name": n, "items": []} for n in DEFAULT_WORLD_CATEGORIES if n != "其他设定"],
            ]
        }

    if isinstance(raw, list):
        cats = []
        for c in raw:
            if isinstance(c, dict) and ("name" in c or "title" in c):
                name = str(c.get("name") or c.get("title") or "未命名")
                items = []
                for it in (c.get("items") or []):
                    if isinstance(it, dict):
                        items.append({
                            "title": str(it.get("title") or it.get("key") or ""),
                            "content": str(it.get("content") or it.get("value") or ""),
                        })
                cats.append({"name": name, "items": items})
        return {"categories": cats or [{"name": n, "items": []} for n in DEFAULT_WORLD_CATEGORIES]}

    return {"categories": [{"name": n, "items": []} for n in DEFAULT_WORLD_CATEGORIES]}


def normalize_memory_fields(
    outline: Any = None,
    storyline: Any = None,
    characters: Any = None,
    world_building: Any = None,
    writing_style: Any = None,
    key_points: Any = None,
    notes: Any = None,
) -> Dict[str, Any]:
    return {
        "outline": outline if isinstance(outline, list) else (outline or []),
        "storyline": normalize_storyline(storyline),
        "characters": normalize_characters(characters),
        "world_building": normalize_world_building(world_building),
        "writing_style": writing_style or "",
        "key_points": normalize_key_points(key_points),
        "notes": notes or "",
    }


def memory_orm_to_dict(memory) -> Dict[str, Any]:
    fields = normalize_memory_fields(
        outline=memory.outline,
        storyline=memory.storyline,
        characters=memory.characters,
        world_building=memory.world_building,
        writing_style=memory.writing_style,
        key_points=memory.key_points,
        notes=memory.notes,
    )
    return {
        "id": memory.id,
        "project_id": memory.project_id,
        "updated_at": memory.updated_at,
        **fields,
    }

File: frontend/src/utils/memoryNormalize.ts
/** 项目设定结构化数据规范化（兼容旧格式） */

export interface StoryStage {
  title: string
  summary: string
}

export interface StorylineData {
  summary: string
  stages: StoryStage[]
}

export interface KeyPointItem {
  title: string
  summary: string
}

export interface WorldItem {
  title: string
  content: string
}

export interface WorldCategory {
  name: string
  items: WorldItem[]
}

export interface WorldBuildingData {
  categories: WorldCategory[]
}

export interface CharacterCard {
  name: string
  description: string
  personality?: string
  background?: string
  goals?: string
  role?: string
  avatar?: string
  color?: string
}

export const DEFAULT_WORLD_CATEGORIES = [
  '时代背景',
  '地理环境',
  '力量/规则',
  '社会结构',
  '文化习俗',
  '其他设定'
]

export const CHAR_ROLES = ['主角', '重要配角', '反派', '配角', '其他']

export const CHAR_COLORS = [
  '#c4a574',
  '#7a9e7e',
  '#6b8cae',
  '#b87a7a',
  '#8b7aad',
  '#c49a6c',
  '#5a8a8a',
  '#a67c52'
]

export function normalizeStoryline(raw: any): StorylineData {
  if (raw == null || raw === '') return { summary: '', stages: [] }
  if (typeof raw === 'string') {
    const s = raw.trim()
    if (s.startsWith('{')) {
      try {
        return normalizeStoryline(JSON.parse(s))
      } catch {
        return { summary: raw, stages: [] }
      }
    }
    return { summary: raw, stages: [] }
  }
  if (Array.isArray(raw)) {
    return { summary: '', stages: raw.map(normalizeStage) }
  }
  if (typeof raw === 'object') {
    const stages = raw.stages || raw.nodes || []
    return {
      summary: String(raw.summary || raw.overview || ''),
      stages: Array.isArray(stages) ? stages.map(normalizeStage) : []
    }
  }
  return { summary: '', stages: [] }
}

function normalizeStage(item: any): StoryStage {
  if (typeof item === 'string') return { title: item.slice(0, 40), summary: item }
  if (item && typeof item === 'object') {
    return {
      title: String(item.title || item.name || ''),
      summary: String(item.summary || item.description || item.content || '')
    }
  }
  return { title: '', summary: '' }
}

export function normalizeKeyPoints(raw: any): KeyPointItem[] {
  if (!Array.isArray(raw)) return []
  return raw.map((item) => {
    if (typeof item === 'string') {
      return { title: item.length > 40 ? item.slice(0, 40) : item, summary: item }
    }
    if (item && typeof item === 'object') {
      const summary = String(item.summary || item.description || item.content || item.point || '')
      const title = String(item.title || item.name || '') || summary.slice(0, 40)
      return { title, summary }
    }
    return { title: '', summary: '' }
  })
}

export function normalizeCharacter(raw: any, index = 0): CharacterCard {
  if (!raw || typeof raw !== 'object') {
    return {
      name: '',
      description: '',
      personality: '',
      background: '',
      goals: '',
      role: '配角',
      avatar: '',
      color: CHAR_COLORS[index % CHAR_COLORS.length]
    }
  }
  return {
    name: String(raw.name || ''),
    description: String(raw.description || ''),
    personality: String(raw.personality || ''),
    background: String(raw.background || ''),
    goals: String(raw.goals || ''),
    role: String(raw.role || '配角'),
    avatar: String(raw.avatar || ''),
    color: String(raw.color || CHAR_COLORS[index % CHAR_COLORS.length])
  }
}

export function normalizeCharacters(raw: any): CharacterCard[] {
  if (!Array.isArray(raw)) return []
  return raw.map((c, i) => normalizeCharacter(c, i))
}

export function normalizeWorldBuilding(raw: any): WorldBuildingData {
  if (!raw) {
    return { categories: DEFAULT_WORLD_CATEGORIES.map((name) => ({ name, items: [] })) }
  }
  if (typeof raw === 'string') {
    try {
      return normalizeWorldBuilding(JSON.parse(raw))
    } catch {
      return { categories: DEFAULT_WORLD_CATEGORIES.map((name) => ({ name, items: [] })) }
    }
  }
  if (raw.categories && Array.isArray(raw.categories)) {
    const cats = raw.categories
      .filter((c: any) => c && typeof c === 'object')
      .map((c: any) => ({
        name: String(c.name || c.title || '未命名分类'),
        items: Array.isArray(c.items)
          ? c.items.map((it: any) => {
              if (typeof it === 'string') return { title: it.slice(0, 30), content: it }
              return {
                title: String(it?.title || it?.key || it?.name || ''),
                content: String(it?.content || it?.value || it?.description || '')
              }
            })
          : []
      }))
    return {
      categories: cats.length
        ? cats
        : DEFAULT_WORLD_CATEGORIES.map((name) => ({ name, items: [] }))
    }
  }
  // 旧扁平 dict
  if (typeof raw === 'object' && !Array.isArray(raw)) {
    const items = Object.entries(raw)
      .filter(([k]) => k !== 'categories')
      .map(([k, v]) => ({ title: String(k), content: v == null ? '' : String(v) }))
    return {
      categories: [
        { name: '综合设定', items },
        ...DEFAULT_WORLD_CATEGORIES
          .filter((n) => n !== '其他设定')
          .map((name) => ({ name, items: [] as WorldItem[] }))
      ]
    }
  }
  return { categories: DEFAULT_WORLD_CATEGORIES.map((name) => ({ name, items: [] })) }
}

export function charInitial(name: string) {
  const n = (name || '').trim()
  return n ? n[0]!.toUpperCase() : '?'
}

export function truncateText(text: string, max = 48) {
  const t = (text || '').replace(/\s+/g, ' ').trim()
  if (t.length <= max) return t
  return t.slice(0, max) + '…'
}
